icu4c/source/tools/upropdump/upropdump.cpp

```diff
@@ -0,0 +1,178 @@
+// © 2016 and later: Unicode, Inc. and others.
+// License & terms of use: http://www.unicode.org/copyright.html
+
+#include <iostream>
+#include "toolutil.h"
+#include "uoptions.h"
+#include "cmemory.h"
+#include "charstr.h"
+#include "unicode/uchar.h"
+#include "unicode/errorcode.h"
+#include "unicode/uniset.h"
+#include "unicode/putil.h"
+#include "unicode/umutablecptrie.h"
+#include "writesrc.h"
+
+U_NAMESPACE_USE
+
+/*
+ * Global - verbosity
+ */
+UBool VERBOSE = FALSE;
+UBool QUIET = FALSE;
+
+UBool haveCopyright = TRUE;
+UCPTrieType trieType = UCPTRIE_TYPE_SMALL;
+
+void handleError(const ErrorCode& status, const char* context) {
+    if (status.isFailure()) {
+        std::cerr << "Error: " << context << ": " << status.errorName() << std::endl;
+        exit(status.get());
+    }
+}
+
+void dumpBinaryProperty(UProperty uproperty, FILE* f) {
+    ErrorCode status;
+    const char* fullPropName = u_getPropertyName(uproperty, U_LONG_PROPERTY_NAME);
+    const char* shortPropName = u_getPropertyName(uproperty, U_SHORT_PROPERTY_NAME);
+    const USet* uset = u_getBinaryPropertySet(uproperty, status);
+    handleError(status, fullPropName);
+
+    fputs("[unicode_set.data]\n", f);
+    fprintf(f, "long_name = \"%s\"\n", fullPropName);
+    usrc_writeUnicodeSet(f, shortPropName, uset, UPRV_TARGET_SYNTAX_TOML);
+}
+
+void dumpEnumeratedProperty(UProperty uproperty, FILE* f) {
+    ErrorCode status;
+    const char* fullPropName = u_getPropertyName(uproperty, U_LONG_PROPERTY_NAME);
+    const char* shortPropName = u_getPropertyName(uproperty, U_SHORT_PROPERTY_NAME);
+    const UCPMap* umap = u_getIntPropertyMap(uproperty, status);
+    handleError(status, fullPropName);
+
+    fputs("[code_point_map.data]\n", f);
```

Comment:
suggestion: I think "code point map" is the interface, while code point trie and inversion map are specific concrete implementations / representations of that interface. Since this header `[code_point_map.data]` is a sibling to the header `[code_point_trie.struct]`, I think it should be renamed to `[inversion_map.data]`

Reply:
I would use `[binary_property.data]` and `[enum_property.data]` -- naming it for what it is rather than what data structure we use.

Reply:
I changed the schema a bit.  Now we have:

```toml
[[enum_property]]
long_name = "General_Category"
short_name = "gc"
# Code points `a` through `b` have value `v`, corresponding to `name`.
ranges = [
  {a=0x0, b=0x1f, v=15, name="Cc"},
  {a=0x20, b=0x20, v=12, name="Zs"},
]

[enum_property.code_point_trie]
index = []
data_8 = []
indexLength = 3365
# ...
```

This becomes a file with a single property, `enum_property`, which is an array of objects with four fields: short_name, long_name, ranges, and code_point_trie.

This seems mildly useful because it could mean you could concatenate these files together and get a well-formed array of properties out of it.

CC @iainireland

06859bf0b283a708a98c1af2ccb97083e54f3019